src/messaging/ExchangeMgr.cpp

```diff
@@ -296,6 +298,50 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
     }
 }
 
+ChannelHandle ExchangeManager::EstablishChannel(const ChannelBuilder & builder, ChannelDelegate * delegate)
+{
+    ChannelContext * channelContext = nullptr;
+
+    // Find an existing Channel matching the builder
+    mChannelContexts.ForEachActiveObject([&](ChannelContext * context) {
```

Comment:
I believe such lambdas use heap (needed to store captures). We try to not use heap directly but rather use platform::MemoryAlloc.

Maybe we could provide an iterator in channel contexts instead of a callback-based interface.

Reply:
No, C++ lambdas doesn't use heap. it will create a temporary struct object on the stack, stores all captured variables in the object (in this case, the references). then pass the structure to the template.

The compiler is able to decide is it worth to inline the lambda or not. When the lambda is inline, it is exactly same as manually iterate. The lambda call back is better in all circumstance.

Reply:
any captures in lambdas were banned in a previous project I worked for because of heap concerns so I remained with that mental model. But now I actually searched it up and apparently my mental model is wrong (e.g. http://www.elbeno.com/blog/?p=1068#:~:text=Lambdas%20don't%20(intrinsically),functor%20optimization%20will%20probably%20apply.)

Thank you for clarifying.